src/Microsoft.ComponentDetection.Orchestrator/Minimatch.cs

```diff
@@ -0,0 +1,843 @@
+                        }
+                    }
+
+                    return true;
+                }
+
+                // ok, let's see if we can swallow whatever we can.
+                for (var i = segmentIndex; i < segments.Count; i++)
+                {
+                    var swallowee = segments[i];
+
+                    // XXX remove this slice.  Just pass the start index.
+                    if (this.MatchOne(segments.Skip(i).ToList(), patterns.Skip(innerPatternIndex).ToList()))
+                    {
+                        // found a match.
+                        return true;
+                    }
+
+                    // can't swallow "." or ".." ever.
+                    if (swallowee.StartsWith(".") || swallowee == "..")
+                    {
+                        break;
+                    }
+
+                    // ** swallows a segment, and continue.
+                }
+
+                return false;
+            }
+
+            // something other than **
+            // non-magic patterns just have to match exactly
+            // patterns with magic have been turned into regexps.
+            if (!pattern.Match(file, this.ignoreCase))
+            {
+                return false;
+            }
+        }
+
+        // Note: ending in / means that we'll get a final string.Empty
+        // at the end of the pattern.  This can only match a
+        // corresponding string.Empty at the end of the file.
+        // If the file ends in /, then it can only match a
+        // a pattern that ends in /, unless the pattern just
+        // doesn't have any more for it. But, a/b/ should *not*
+        // match "a/b/*", even though string.Empty matches against the
+        // [^/]*? pattern, except in partial mode, where it might
+        // simply not be reached yet.
+        // However, a/b/ should still satisfy a/*
+
+        // now either we fell off the end of the pattern, or we're done.
+        if (segmentIndex == segments.Count && patternIndex == patterns.Count)
+        {
+            // ran out of pattern and filename at the same time.
+            // an exact hit!
+            return true;
+        }
+
+        if (segmentIndex == segments.Count)
+        {
+            // ran out of file, but still had pattern left.
+            // this is ok if we're doing the match as part of
+            // a glob fs traversal.
+            return false;
+        }
+
+        if (patternIndex == patterns.Count)
+        {
+            // ran out of pattern, still have file left.
+            // this is only acceptable if we're on the very last
+            // empty segment of a file with a trailing slash.
+            // a/* should match a/b/
+            var emptyFileEnd = segmentIndex == segments.Count - 1 && segments[segmentIndex].Length == 0;
+            return emptyFileEnd;
+        }
+
+        // should be unreachable.
+        throw new InvalidOperationException("wtf?");
```

Comment:
nit: I know it comes from the original source, but can we give this a message that is more clear?